src/System.Linq/src/System/Linq/Select.cs

```diff
@@ -167,14 +175,15 @@ public override Iterator<TResult> Clone()
 
             public override bool MoveNext()
             {
-                if (_state == 1 && _index < _source.Length)
+                if (_state == 0 | _state == _source.Length + 1)
```

Comment:
I used `|` to be consistent with other parts of the LINQ codebase where bitwise instead of logical operators are being used, e.g. [here](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Select.cs#L713). Since it will be very rare for `_state` to be 0 (only happens when someone casts the result of `Select` to an `IEnumerator`) this will save us a branch.